### PR TITLE
Switch the index to basic messaging

### PR DESCRIPTION
**Pull request by tobim**
This PR removes the use of CAF streaming from the index component, which was the last that still used it.
The data flow is now modeled with simple sends, so data is not acked. This was done intentionally because it is simpler, and acks are not implemented in the pipelines layer yet either.

### CODE REVIEW

**Review by lava — APPROVED** [review]
Insane to see in one place how much complexity the streaming part was generating.

File: libtenzir/src/index.cpp [hunk 1]
```diff
@@ -1785,34 +1708,10 @@ index(index_actor::stateful_pointer<index_state> self,
     [self](atom::flush) -> caf::result<void> {
       TENZIR_DEBUG("{} got a flush request from {}", *self,
                    self->current_sender());
-      // If we've got nothing to flush we can just exit immediately.
-      if (self->state.active_partitions.empty())
+      if (self->state.active_partitions.empty()) {
         return {};
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/index.cpp:1712]: Yes, but I find it cleaner to not create a response promise when we don't need to delegate the request.

File: libtenzir/src/importer.cpp [hunk 2]
```diff
@@ -158,17 +79,11 @@ importer(importer_actor::stateful_pointer<importer_state> self,
       std::filesystem::exists(dir / "current_id_block", ec)) {
     std::filesystem::remove(dir / "current_id_block", ec);
   }
-  namespace defs = defaults;
   self->set_exit_handler([=](const caf::exit_msg& msg) {
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/importer.cpp:82]: The index flushes automatically when it shuts down.

File: libtenzir/builtins/operators/import.cpp [hunk 3]
```diff
@@ -88,10 +88,10 @@ class import_operator final : public crtp_operator<import_operator> {
             std::chrono::duration<double, std::chrono::seconds::period>>(
             elapsed)
             .count();
-    TENZIR_DEBUG("imported {} events in {}{}", total_events, data{elapsed},
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/builtins/operators/import.cpp]: Leftover.

File: libtenzir/src/index.cpp [hunk 4]
```diff
@@ -69,6 +67,7 @@
 
 // clang-format off
 //
+// ### OBSOLETE!!!!
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/index.cpp]: Cleaned it up instead.

File: libtenzir/src/index.cpp [hunk 5]
```diff
@@ -1167,12 +1088,20 @@ index(index_actor::stateful_pointer<index_state> self,
...
-    for (auto&& [rp, _] : std::exchange(self->state.delayed_queries, {}))
+    TENZIR_VERBOSE("{} received EXIT from {} with reason: {}", *self,
+                   msg.source, msg.reason);
+    for (auto&& [rp, _] : std::exchange(self->state.delayed_queries, {})) {
       rp.deliver(msg.reason);
-    // Flush buffered batches and end stream.
-    self->state.stage->stop();
+    }
+    self->state.shutting_down = true;
+    self->request(static_cast<index_actor>(self), caf::infinite, atom::flush_v)
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/index.cpp]: Added a pretty graceful 120 seconds.

File: libtenzir/src/partition_transformer.cpp [hunk 6]
```diff
@@ -445,15 +429,10 @@ auto partition_transformer(
...
           return {};
         }
-        self->monitor(builder_and_header->store_builder);
+        partition_data.builder = builder_and_header->store_builder;
+        self->monitor(partition_data.builder);
         ++self->state.stores_launched;
         partition_data.store_header = builder_and_header->header;
-        // Empirically adding the outbound path and pushing data to it
-        // need to be separated by a continuation, although I'm not
-        // completely sure why.
```
[the remark this answers is not in the document]
Comment (tobim) [reply libtenzir/src/partition_transformer.cpp:453]: We could, but while the current method is not the most efficient, it works, and we can spend our time on more impactful things.

File: libtenzir/builtins/components/metrics_collector.cpp [hunk 7]
```diff
@@ -57,6 +58,9 @@ struct metrics_collector_state {
   std::vector<instance> instances = {};
 
   auto setup() -> caf::expected<void> {
+    if (caf::get_or(self->system().config(), "tenzir.enable-metrics", true)) {
+      return {};
+    }
```
Comment (dominiklohmann) [inline libtenzir/builtins/components/metrics_collector.cpp:63]: This disables instead of enabling the metric if the option is set.

In any case, as we discussed this morning, please undo this change, and remove the option in its entirety.

File: libtenzir/src/partition_transformer.cpp [hunk 8]
```diff
@@ -493,7 +457,16 @@ auto partition_transformer(
...
       }
-      detail::shutdown_stream_stage(self->state.stage);
+      for (auto& [_, partition_data] : self->state.data) {
+        self->request(partition_data.builder, caf::infinite, atom::persist_v)
+          .then(
+            [](resource&) {
+              // nop
+            },
+            [](caf::error&) {
+              // nop
```
Comment (dominiklohmann) [inline libtenzir/src/partition_transformer.cpp:467]: This should be handled. At least a warning, if not a hard error. We have a data loss here.

File: libtenzir/src/partition_transformer.cpp [hunk 9]
```diff
@@ -493,7 +457,16 @@ auto partition_transformer(
           data.indexer_chunks.emplace_back(qf.name(), chunk);
         }
       }
-      detail::shutdown_stream_stage(self->state.stage);
+      for (auto& [_, partition_data] : self->state.data) {
+        self->request(partition_data.builder, caf::infinite, atom::persist_v)
+          .then(
+            [](resource&) {
+              // nop
```
Comment (dominiklohmann) [inline libtenzir/src/partition_transformer.cpp:464]: This should deliver a response promise for the parent handler once they're all done, or at least have a TODO comment for that (with a follow-up story).

File: libtenzir/src/active_partition.cpp [hunk 10]
```diff
@@ -475,16 +333,8 @@ active_partition_actor::behavior_type active_partition(
...
-    // on 'std::vector<caf::actor>' only. That should probably be generalized
-    // in the future.
-    auto indexers = std::vector<caf::actor>{};
-    indexers.reserve(self->state.indexers.size());
-    auto copy = std::exchange(self->state.indexers, {});
-    for ([[maybe_unused]] auto&& [qf, indexer] : std::move(copy))
-      indexers.push_back(caf::actor_cast<caf::actor>(std::move(indexer)));
-    shutdown<policy::parallel>(self, std::move(indexers));
+    TENZIR_DEBUG("{} shuts down after persisting partition state", *self);
+    self->quit();
```
Comment (dominiklohmann) [inline libtenzir/src/active_partition.cpp:337]: We should consider quitting with `msg.reason` here or ideally `diagnostic::error(msg.reason).note("further context").to_error()`.

File: libtenzir/src/store.cpp [hunk 11]
```diff
@@ -281,88 +280,44 @@ default_active_store_actor::behavior_type default_active_store(
...
     },
     [self](atom::persist) -> caf::result<resource> {
-      if (const auto* res = std::get_if<resource>(&self->state.file)) {
-        return *res;
+      if (self->state.erased) {
+        return {};
+      }
+      auto chunk = self->state.store->finish();
+      if (!chunk) {
+        self->quit(std::move(chunk.error()));
```
Comment (dominiklohmann) [inline libtenzir/src/store.cpp:288]: Consider adding some context here, e.g., `diagnostic::error(chunk.error()).note(...).to_error()`.

File: libtenzir/src/store.cpp [hunk 12]
```diff
@@ -281,88 +280,44 @@ default_active_store_actor::behavior_type default_active_store(
...
+          });
+      return rp;
+    },
+    [self](table_slice& slice) {
+      if (self->state.erased) {
+        return;
+      }
+      // TODO: Get rid of the vector.
+      if (auto error = self->state.store->add(std::vector{std::move(slice)})) {
+        self->quit(std::move(error));
```
Comment (dominiklohmann) [inline libtenzir/src/store.cpp:319]: Same here, consider adding some context.

File: libtenzir/src/index.cpp [hunk 13]
```diff
@@ -1167,12 +1089,22 @@ index(index_actor::stateful_pointer<index_state> self,
...
+    self->state.shutting_down = true;
+    self
+      ->request(static_cast<index_actor>(self), std::chrono::seconds{120},
+                atom::flush_v)
+      .then(
+        [self]() {
+          self->quit();
+        },
+        [self](caf::error& err) {
+          self->quit(err);
```
Comment (dominiklohmann) [inline libtenzir/src/index.cpp:1106]: Please add some context here, and provide a special error message if `err == caf::sec::request_timeout` that is very visible.

File: libtenzir/src/index.cpp [hunk 14]
```diff
@@ -1167,12 +1089,22 @@ index(index_actor::stateful_pointer<index_state> self,
...
+                   msg.source, msg.reason);
+    for (auto&& [rp, _] : std::exchange(self->state.delayed_queries, {})) {
       rp.deliver(msg.reason);
-    // Flush buffered batches and end stream.
-    self->state.stage->stop();
+    }
+    self->state.shutting_down = true;
+    self
+      ->request(static_cast<index_actor>(self), std::chrono::seconds{120},
+                atom::flush_v)
```
Comment (dominiklohmann) [inline libtenzir/src/index.cpp:1100]: Let's keep an eye on this one for a while. Consider increasing the timeout to 10 min or so.

File: contrib/tenzir-plugins
Comment (dominiklohmann) [inline contrib/tenzir-plugins:1]: The `context_plugin::snapshot` function can go. Perhaps also `context_plugin::make_query`. Can you write a follow-up story for this?

**Review by dominiklohmann — APPROVED** [review]
Approving the code following our paired review with a few comments. Will do some manual tests for the shutdown logic, rebuilding, and compaction now.

File: contrib/tenzir-plugins
Comment (dominiklohmann) [inline contrib/tenzir-plugins:1]: The lookup changes in this also need a docs change and a changelog entry.
